### PR TITLE
added six task. Complete module 2.3.

### PR DESCRIPTION
**Pull request by streamdp**

### CODE REVIEW

File: src/main/java/com/epam/streamdp/six/AddTrain.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+import static com.epam.streamdp.six.Trains.LEFT;
+import static com.epam.streamdp.six.Trains.RIGHT;
+
+public class AddTrain implements Runnable {
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(AddTrain.class.getName());
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/AddTrain.java]: Please use static import for the class

File: src/main/java/com/epam/streamdp/six/AddTrain.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private static final int MAXIMUM_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN = 1000;
+    private static final int BASE_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN = 1100;
+    private static final String EXCEPTION_MESSAGE = "Exception: ";
+    private static final int HOW_MUCH_TRAINS_WILL_BE_GENERATE = 100;
+    private String[] location = {LEFT, RIGHT};
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/AddTrain.java]: That could be an enum

File: src/main/java/com/epam/streamdp/six/AddTrain.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.epam.streamdp.six;
+
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+import static com.epam.streamdp.six.Trains.LEFT;
+import static com.epam.streamdp.six.Trains.RIGHT;
+
+public class AddTrain implements Runnable {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/AddTrain.java]: A class name should sound like a noun, e.g. TrainGenerator, TrainStarter, etc. Methods names should use verbs: createTrain, startTrain, addTrain, etc

File: src/main/java/com/epam/streamdp/six/AddTrain.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(AddTrain.class.getName());
+    private static final int MAXIMUM_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN = 1000;
+    private static final int BASE_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN = 1100;
+    private static final String EXCEPTION_MESSAGE = "Exception: ";
+    private static final int HOW_MUCH_TRAINS_WILL_BE_GENERATE = 100;
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/AddTrain.java]: HOW_MUCH_TRAINS_WILL_BE_GENERATED

File: src/main/java/com/epam/streamdp/six/AddTrain.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                TimeUnit.MILLISECONDS.sleep(random.nextInt(MAXIMUM_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN) + BASE_VALUE_OF_DELAY_FOR_GENERATE_NEXT_TRAIN);
+            }
+            tunnel.setCycle(false);
+        } catch (Exception ex) {
+            logger.log(Level.SEVERE, EXCEPTION_MESSAGE, ex);
+        }
+    }
+
+    private String makeLogSting(Trains train) {
+        return "The train №" + train.trainNumber + " drive up to the tunnel №" + train.tunnelNumber + " from " + train.location + " side";
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/AddTrain.java]: It is preferable to use String.format() for concatination complex strings

File: src/main/java/com/epam/streamdp/six/Balancer.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import static com.epam.streamdp.six.MainTask.ONE;
+import static com.epam.streamdp.six.MainTask.TWO;
+import static com.epam.streamdp.six.Trains.LEFT;
+import static com.epam.streamdp.six.Trains.RIGHT;
+
+public class Balancer implements Runnable {
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(Balancer.class.getName());
+    private static final String EXCEPTION_MESSAGE = "Exception: ";
+    private static final String MESSAGE_FROM_ONE_TO_TWO = "Event: \"The train №{0} was moved from the queue to the tunnel ONE to the queue to the tunnel TWO\"";
+    private static final String MESSAGE_FROM_TWO_TO_ONE = "Event: \"The train №{0} was moved from the queue to the tunnel TWO to the queue to the tunnel ONE\"";
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Balancer.java]: You can create a string pattern with placeholders and just pass needed parameters in runtime.

```
private static final String TRAIN_HAS_MOVED_MESSAGE = "Event: 'The train №{0} was moved from the queue to the tunnel %s to the queue to the tunnel %s'";

String.format(TRAIN_HAS_MOVED_MESSAGE, ONE, TWO);
String.format(TRAIN_HAS_MOVED_MESSAGE, TWO, ONE);
```

File: src/main/java/com/epam/streamdp/six/Balancer.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public Trains getTrainsBalancingByTime(Trains train, String side) {
+        train = getTotalQueueMovementTime(tunnelOne, side) > getTotalQueueMovementTime(tunnelTwo, side)
+                ? moveTrainFromTunnelNumToTunnelNum(train, tunnelOne, tunnelTwo, TWO, MESSAGE_FROM_ONE_TO_TWO)
+                : moveTrainFromTunnelNumToTunnelNum(train, tunnelTwo, tunnelOne, ONE, MESSAGE_FROM_TWO_TO_ONE);
+        return train;
+    }
+
+    public boolean verifyBalancingByTimeAbility(Tunnels tunnelOne, Tunnels tunnelTwo, String side) {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Balancer.java]: Methods that return a boolean value have a special naming convention and they should start with for example 'has', 'is', 'can', 'able', etc. verifyBalancingByTimeAbility say nothing for me either it returns true or false

File: src/main/java/com/epam/streamdp/six/RemoveTrain.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package com.epam.streamdp.six;
+
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+public class RemoveTrain implements Runnable {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/RemoveTrain.java]: That sounds like a method

File: src/main/java/com/epam/streamdp/six/RemoveTrain.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(RemoveTrain.class.getName());
+    private static final String EXCEPTION_MESSAGE = "Exception: ";
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/RemoveTrain.java]: 4 classes contain the constant. Probably it's time to stare it in one place and reuse it?

File: src/main/java/com/epam/streamdp/six/Trains.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
+package com.epam.streamdp.six;
+
+public class Trains {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Trains.java]: Usually, we don't use plural form for class names. I suggest to use the Train class and create an array or list to store several trains

File: src/main/java/com/epam/streamdp/six/Trains.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        this.trainNumber = trainNumber;
+        this.location = location;
+        this.tunnelNumber = tunnelNumber;
+    }
+
+    public void setTunnelNumber(int tunnelNumber) {
+        this.tunnelNumber = tunnelNumber;
+    }
+
+    public int getHowFastWillTheTrainTunnelGo() {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Trains.java]: '...TrainTunnelGo' does a tunnel go?
words order is important: getHowFastWillTheTrainGoInsideATunnel has totally different meaning

File: src/main/java/com/epam/streamdp/six/Tunnels.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
+package com.epam.streamdp.six;
+
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.logging.Level;
+
+public class Tunnels implements Runnable {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Tunnels.java]: Tunnel

File: src/test/java/com/epam/streamdp/six/AddTrainTest.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import static com.epam.streamdp.six.Trains.LEFT;
+import static com.epam.streamdp.six.Trains.RIGHT;
+
+public class AddTrainTest {
+
+    @Test(description = "Verify method creating the train")
+    public void generateSomeTrain() {
+        SoftAssert softAssertion = new SoftAssert();
+        Trains train = new AddTrain().generateSomeTrain(1, 1);
+        softAssertion.assertTrue(train != null, "Error generating train, train is NULL.");
```
Comment (yury-barsukou) [inline src/test/java/com/epam/streamdp/six/AddTrainTest.java]: Use assertNotNull

File: src/test/java/com/epam/streamdp/six/AddTrainTest.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        softAssertion.assertTrue(train.location.equals(LEFT) || train.location.equals(RIGHT), "Train location must be LEFT of RIGHT");
+        softAssertion.assertTrue(train.trainNumber > 0, "Train number must be > 0.");
+        softAssertion.assertTrue(train.tunnelNumber >= 0, "Train number can not be negative.");
+        softAssertion.assertTrue(train.countOfRailWagons > 0, "Train number must be > 0.");
+        softAssertion.assertAll();
```
Comment (yury-barsukou) [inline src/test/java/com/epam/streamdp/six/AddTrainTest.java]: Don't overdo with using SoftAssert. Only one conception should be checked in a group. Don't mix all checks inside, better split it onto different tests if it's possible

File: src/test/java/com/epam/streamdp/lerning/hellowebdriver/HelloWebDriver.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package com.epam.streamdp.lerning.hellowebdriver;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
+
+public class HelloWebDriver {
```
Comment (yury-barsukou) [inline src/test/java/com/epam/streamdp/lerning/hellowebdriver/HelloWebDriver.java]: How did the code appear here? Is it from another task? You should make changes in the particular branch

File: src/test/java/com/epam/streamdp/lerning/waits/WebDriverSyncDemo.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.util.List;
+
+public class WebDriverSyncDemo {
```
Comment (yury-barsukou) [inline src/test/java/com/epam/streamdp/lerning/waits/WebDriverSyncDemo.java]: The same

File: src/main/java/com/epam/streamdp/six/Balancer.java [hunk 17]
```diff
@@ -41,35 +40,42 @@ public void run() {
...
-        return tunnelOne.getTunnel().stream().filter(train -> train.location.equals(side)).count() > MINIMUM_QUEUE_SIZE_TO_ALLOW_BALANCING_BY_TIME
-                && tunnelTwo.getTunnel().stream().filter(train -> train.location.equals(side)).count() > MINIMUM_QUEUE_SIZE_TO_ALLOW_BALANCING_BY_TIME;
+    public String makeMessage(String from, String to) {
+        return String.format("The train №{0} was moved from the queue to the tunnel %s to the queue to the tunnel %s",
+                from, to);
     }
 
-    public Trains moveTrainFromTunnelNumToTunnelNum(Trains previousTrain, Tunnels tunnelFrom, Tunnels tunnelTo, int tunnelNumber, String message) {
-        Trains train = tunnelFrom.getTunnel().peekLast();
+    public boolean verifyBalancingByTimeAbility(Tunnel tunnelOne, Tunnel tunnelTwo, Locations side) {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Balancer.java]: Please adjust method's name according to my recommendations for "boolean" methods

File: src/main/java/com/epam/streamdp/six/TrainDestroyer.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
+package com.epam.streamdp.six;
+
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+import static com.epam.streamdp.six.Tunnel.EXCEPTION_MESSAGE;
+
+public class TrainDestroyer implements Runnable {
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(TrainDestroyer.class.getName());
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/TrainDestroyer.java]: Get rid of the full class name that includes packages

File: src/main/java/com/epam/streamdp/six/TrainGenerator.java [hunk 19]
```diff
@@ -4,33 +4,30 @@
...
     }
 
     @Override
     public void run() {
         try {
-            for (int i = 1; i <= HOW_MUCH_TRAINS_WILL_BE_GENERATE; i++) {
-                Trains train = generateSomeTrain(1000, tunnel.getTunnelNumber());
-                tunnel.getTunnel().addFirst(train);
+            for (int i = 1; i <= HOW_MUCH_TRAINS_WILL_BE_GENERATED
+                    ; i++) {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/TrainGenerator.java]: Why do you need the line break here?

File: src/main/java/com/epam/streamdp/six/Locations.java [hunk 20]
```diff
@@ -0,0 +1,5 @@
+package com.epam.streamdp.six;
+
+public enum Locations {
```
Comment (yury-barsukou) [inline src/main/java/com/epam/streamdp/six/Locations.java]: Location

File: src/test/java/com/epam/streamdp/lerning/hellowebdriver/HelloWebDriver.java (hunk 15 above)
Comment (streamdp) [reply src/test/java/com/epam/streamdp/lerning/hellowebdriver/HelloWebDriver.java]: I'm sorry it shouldn't have gotten here. I really did it in another brunch.

File: src/main/java/com/epam/streamdp/six/Balancer.java (hunk 17 above)
Comment (streamdp) [reply src/main/java/com/epam/streamdp/six/Balancer.java]: i'm sorry

**Pull request merged**